blockchain/src/codec.rs

```diff
@@ -0,0 +1,388 @@
+use crate::shortid::ShortIDVec;
+use crate::utreexo::Proof;
+use crate::{
+    Block, BlockHeader, BlockID, BlockTx, GetBlock, GetInventory, GetMempoolTxs, Inventory,
+    MempoolTxs, Message,
+};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use readerwriter::{Decodable, Encodable, ReadError, Reader, WriteError, Writer};
+use std::convert::TryFrom;
+use zkvm::{merkle, Hash, Signature, Tx};
+
+#[repr(u8)]
+enum MessageType {
+    Block = 0,
+    GetBlock = 1,
+    Inventory = 2,
+    GetInventory = 3,
+    MempoolTxs = 4,
+    GetMempoolTxs = 5,
+}
+
+impl TryFrom<u8> for MessageType {
+    type Error = ReadError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(MessageType::Block),
+            1 => Ok(MessageType::GetBlock),
+            2 => Ok(MessageType::Inventory),
+            3 => Ok(MessageType::GetInventory),
+            4 => Ok(MessageType::MempoolTxs),
+            5 => Ok(MessageType::GetMempoolTxs),
+            _ => Err(ReadError::Custom(
+                format!("unknown message type: {}", value).into(),
+            )),
+        }
+    }
+}
+
+impl Encodable for BlockHeader {
+    fn encode(&self, w: &mut impl Writer) -> Result<(), WriteError> {
+        w.write_u64(b"version", self.version)?;
+        w.write_u64(b"height", self.height)?;
+        w.write_blockid(b"prev", &self.prev)?;
+        w.write_u64(b"timestamp_ms", self.timestamp_ms)?;
+        w.write_hash(b"txroot", &self.txroot)?;
+        w.write_hash(b"utxoroot", &self.utxoroot)?;
+        w.write_u8_vec(b"ext", &self.ext)?;
+        Ok(())
+    }
+
+    fn encoded_length(&self) -> usize {
+        8 + 8 + 32 + 8 + 32 + 32 + 4 + self.ext.len()
+    }
+}
+
+impl Decodable for BlockHeader {
+    fn decode(buf: &mut impl Reader) -> Result<Self, ReadError> {
+        Ok(BlockHeader {
+            version: buf.read_u64()?,
+            height: buf.read_u64()?,
+            prev: buf.read_blockid()?,
+            timestamp_ms: buf.read_u64()?,
+            txroot: buf.read_hash()?,
+            utxoroot: buf.read_hash()?,
+            ext: buf.read_u8_vec()?,
+        })
+    }
+}
+
+impl Encodable for Inventory {
+    fn encode(&self, w: &mut impl Writer) -> Result<(), WriteError> {
+        w.write_u64(b"version", self.version)?;
+        self.tip.encode(w)?;
+        w.write_signature(&self.tip_signature)?;
+        w.write_u64(b"shortid_nonce", self.shortid_nonce)?;
+        w.write_shortid_vec(b"shortid_list", &self.shortid_list)?;
+        Ok(())
+    }
+
+    fn encoded_length(&self) -> usize {
+        8 + self.tip.encoded_length() + 32 + 8 + 4 + self.shortid_list.len()
```

Comment:
1) signature is 64 bytes long.
2) shortid_list.len() is in items, not bytes. 

Note for later: ShortID should verify that the buffer is divisible by SHORTID_LEN so we don't accidentally have extra bytes that will be re-encoded, but not converted into IDs. Right now there's such possibility (see .map(ShortIDVec) below).

```suggestion
        8 + self.tip.encoded_length() + 64 + 8 + 4 + self.shortid_list.len()*SHORTID_LEN
```

Reply:
Filed here: https://github.com/stellar/slingshot/issues/436

Reply:
Done